stdlib/cache/src/main/ballerina/src/cache/linked_list.bal

```diff
@@ -43,77 +43,94 @@ public type LinkedList object {
+                self.tail = self.head;
+                releaseLock();
+                return;
+            }
+            Node tempNode = node;
+            Node tailNode = <Node>self.tail;
+            tempNode.prev = tailNode;
+            tailNode.next = tempNode;
+            self.tail = tempNode;
+            releaseLock();
         }
-        Node tempNode = node;
-        Node tailNode = <Node>self.tail;
-        tempNode.prev = tailNode;
-        tailNode.next = tempNode;
-        self.tail = tempNode;
     }
 
     # Adds a node to the start of the provided linked list.
     #
     # + node - The node, which should be added to the provided linked list
     public function addFirst(Node node) {
-        if (self.head is ()) {
-            self.head = node;
-            self.tail = self.head;
-            return;
+        if (tryLock()) {
+            if (self.head is ()) {
+                self.head = node;
+                self.tail = self.head;
+                releaseLock();
+                return;
+            }
+            Node tempNode = node;
+            Node headNode = <Node>self.head;
+            tempNode.next = headNode;
+            headNode.prev = tempNode;
+            self.head = tempNode;
+            releaseLock();
         }
-        Node tempNode = node;
-        Node headNode = <Node>self.head;
-        tempNode.next = headNode;
-        headNode.prev = tempNode;
-        self.head = tempNode;
     }
 
     # Removes a node from the provided linked list.
     #
     # + node - The node, which should be removed from the provided linked list
     public function remove(Node node) {
-        if (tryLock()) {
-            if (node.prev is ()) {
-                self.head = node.next;
-            } else {
-                Node prev = <Node>node.prev;
-                prev.next = node.next;
-            }
-            if (node.next is ()) {
-                self.tail = node.prev;
-            } else {
-                Node next = <Node>node.next;
-                next.prev = node.prev;
-            }
-            node.next = ();
-            node.prev = ();
-            releaseLock();
+        if (node.prev is ()) {
+            self.head = node.next;
+        } else {
+            Node prev = <Node>node.prev;
+            prev.next = node.next;
+        }
+        if (node.next is ()) {
+            self.tail = node.prev;
+        } else {
+            Node next = <Node>node.next;
+            next.prev = node.prev;
         }
+        node.next = ();
+        node.prev = ();
     }
 
     # Removes the last node from the provided linked list.
     #
+    # + node - The node, which should be removed from the provided linked list
     # + return - Last node of the provided linked list or `()` if the last node is empty
-    public function removeLast() returns Node? {
-        if (self.tail is ()) {
-            return ();
+    public function removeLast(Node? node = ()) returns Node? {
+        if (tryLock()) {
```

Comment:
We don't need to `tryLock()` here. We need to call `remove` function inside of this function.